### PR TITLE
[ZEPPELIN-1142] Zeppelin allows two users to simultaneously edit the notebook permissions

### PR DESCRIPTION
**Pull request by kavinkumarks**
### What is this PR for?
This is about storing the owner information on creating a note so when the same user tries to edit the permissions of the note he could do it successfully.

### What type of PR is it?
Improvement

### Todos
NA

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-1142

### How should this be tested?
Check the below cases via the web application or REST API calls and the owner information should be persisted properly.
* Create note
* Clone note
* Import note

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java [hunk 1]
```diff
@@ -210,6 +211,21 @@ public void testPersist() throws IOException, SchedulerException, RepositoryExce
   }
 
   @Test
+  public void testCreateNoteWithSubject() throws IOException, SchedulerException, RepositoryException {
+    AuthenticationInfo subject = new AuthenticationInfo("user1");
+    Note note = notebook.createNote(subject);
+
+    Notebook notebook = new Notebook(
```
Comment (khalidhuseynov) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java]: why creating new `notebook` instance here? can just reused previous one, isn't it

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java [hunk 2]
```diff
@@ -396,6 +424,18 @@ public void testCloneNote() throws IOException, CloneNotSupportedException,
     assertEquals(cp.getId(), p.getId());
     assertEquals(cp.text, p.text);
     assertEquals(cp.getResult().message(), p.getResult().message());
+
+    // Verify clone note with subject
+    AuthenticationInfo subject = new AuthenticationInfo("user1");
+    Note cloneNote2 = notebook.cloneNote(note.getId(), "clone note2", subject);
+    Notebook notebook = new Notebook(
```
Comment (khalidhuseynov) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java]: same here with `notebook` instance

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java [hunk 3]
```diff
@@ -374,11 +390,23 @@ public void testExportAndImportNote() throws IOException, CloneNotSupportedExcep
     assertEquals(p.getId(), p2.getId());
     assertEquals(p.text, p2.text);
     assertEquals(p.getResult().message(), p2.getResult().message());
+
+    // Verify import note with subject
+    AuthenticationInfo subject = new AuthenticationInfo("user1");
+    Note importedNote2 = notebook.importNote(exportedNoteJson, "Title2", subject);
+    Notebook notebook = new Notebook(
```
Comment (khalidhuseynov) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java]: here as well

**Comment by khalidhuseynov** [comment]
@kavinkumarks thanks for improvement and it makes sense indeed to assign owner when creating note. just need to address above minor comments as well as restart CI, some profiles are failing.

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java (hunk 1 above)
Comment (kavinkumarks) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java]: The `notebook` instance is different from `note` where the former holds the collection of the latter.

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java (hunk 2 above)
Comment (kavinkumarks) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java]: Please refer to my first comment about this.

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java (hunk 3 above)
Comment (kavinkumarks) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java]: Please refer to my first comment about this.

**Comment by kavinkumarks** [comment]
@khalidhuseynov thanks for the review! I have replied to the comments and the CI build is in progress.

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java (hunk 1 above)
Comment (khalidhuseynov) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java]: yeah that's true, but the question is why you instantiate `notebook` with `new Notebook` on line 218 if you already used different instance of it to create `note` on line 216.  you can reuse same instance and just delete instantiation at line 218 , isn't it?

Comment (kavinkumarks) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java]: Yeah that was correct.Sorry for missing them out. I have made the changes and pushed them.

**Comment by kavinkumarks** [comment]
Reopening to trigger the CI build.

**Comment by kavinkumarks** [comment]
Found an build test failure, working on it.

**Comment by kavinkumarks** [comment]
Fixed the build test failure and the CI build is in progress.

**Comment by kavinkumarks** [comment]
The CI build is green.

**Comment by khalidhuseynov** [comment]
LGTM

**Comment by kavinkumarks** [comment]
Thanks @khalidhuseynov ! Can we get this merged?

-Kavin
MailTo: kavin.kumar@imaginea.com

**Comment by khalidhuseynov** [comment]
@kavinkumarks i believe some of the committers or pmcs will take a look into this pr before merging

**Comment by kavinkumarks** [comment]
Could someone else have a look at this and merge if everything is okay?

Thanks,
Kavin
MailTo: kavin.kumar@imaginea.com

**Comment by Leemoonsoo** [comment]
Thanks @kavinkumarks for the contribution.
This changes not only resolving race condition described in ZEPPELIN-1142 but also changes user experience. i.e.

Before, owner is not defined when creating note.
After, owner is set to current user when creating note. (when authentication is enabled)

Especially with code changes such as https://github.com/apache/zeppelin/pull/1330, user will experience bigger behavior changes.

But i think having current user as a owner when creating note is natural behavior that most people can think, even though it brings some UX changes.

LGTM and merge if there're no further discussions.

**Comment by kavinkumarks** [comment]
Thanks @Leemoonsoo for reviewing this! Could you please merge this?

- Kavin
MailTo: kavin.kumar@imaginea.com

**Pull request closed without merging**